R/layers.R

```diff
@@ -237,12 +241,120 @@ addMarkers = function(
+#'   (relative to its top left corner, i.e. the top left corner means
+#'   \code{iconAnchorX = 0} and \code{iconAnchorY = 0)}, and the icon will be
+#'   aligned so that this point is at the marker's geographical location
+#' @param shadowUrl the URL to the icon shadow image
+#' @param shadowRetinaUrl the URL to the retina sized version of the icon shadow
+#'   image
+#' @param shadowSizeX,shadownSizeY size of the shadow image in pixels
+#' @param shadowAnchorX,shadowAnchorY the coordinates of the "tip" of the shadow
+#' @param popupAnchorX,popupAnchorY the coordinates of the point from which
+#'   popups will "open", relative to the icon anchor
+#' @param className a custom class name to assign to both icon and shadow images
+#' @param embed whether to base64 encode local image files
+#' @note The argument \code{embed = TRUE} can be useful when you render the map
+#'   in a static HTML document and you want it to be self-contained (i.e. no
+#'   external image dependencies). If you are sure that the icon images will
+#'   always be available along with the HTML document, it is not necessary to
+#'   encode them. When using the icons in Shiny, you may (pre-)render the icon
+#'   images under the \file{www} directory of the app, and use them without
+#'   base64 encoding them, e.g. you can use an icon \file{www/foo.png} by
+#'   \code{iconList(iconUrl = 'foo.png', embed = FALSE)} (note there is no
+#'   \samp{www} prefix in the icon URL).
+#' @return A list of icon data that can be passed to the \code{icon} argument of
+#'   \code{\link{addMarkers}()}.
+#' @export
+#' @example inst/examples/iconList.R
+iconList = function(
+  iconUrl = NULL, iconRetinaUrl = NULL, iconSizeX = NULL, iconSizeY = NULL,
+  iconAnchorX = NULL, iconAnchorY = NULL, shadowUrl = NULL, shadowRetinaUrl = NULL,
+  shadowSizeX = NULL, shadowSizeY = NULL, shadowAnchorX = NULL, shadowAnchorY = NULL,
+  popupAnchorX = NULL, popupAnchorY = NULL, className = NULL, embed = TRUE
+) {
+  op = options(stringsAsFactors = FALSE); on.exit(options(op))
+  attrs = filterNULL(list(
+    iconUrl = iconUrl, iconRetinaUrl = iconRetinaUrl,
+    iconSizeX = iconSizeX, iconSizeY = iconSizeY,
+    iconAnchorX = iconAnchorX, iconAnchorY = iconAnchorY,
+    shadowUrl = shadowUrl, shadowRetinaUrl = shadowRetinaUrl,
+    shadowSizeX = shadowSizeX, shadowSizeY = shadowSizeY,
+    shadowAnchorX = shadowAnchorX, shadowAnchorY = shadowAnchorY,
+    popupAnchorX = popupAnchorX, popupAnchorY = popupAnchorY,
+    className = className
+  ))
+  attrs = as.data.frame(do.call(cbind, attrs))
+  group = apply(as.matrix(attrs), 1, paste, collapse = '\r') # one row to one value
+  group = as.integer(factor(group, unique(group)))  # map rows to groups
+  attrs = attrs[!duplicated(group), , drop = FALSE] # dedup
+  structure(
+    lapply(seq_len(nrow(attrs)), function(i) {
+      L.icon(iconData(attrs[i, , drop = FALSE]), embed = embed)
+    }),
+    iconGroup = group
+  )
```

Comment:
Yeah, this does look complicated. I thought quite a while about it. The two major issues that make it tricky to recycle icon arguments are:

1. The `icon` argument of `L.marker()` is at the same level as other marker options such as `clickable` and `draggable`, and `iconUrl`/`iconSize`/... are arguments of `L.icon()`. I'm not sure if you want to put all icon arguments into the `options` argument of `addMarkers()`, which will raise them to the same level as `clickable` after `dataframe.cbind(options)` in the client.
2. Recycling of icon arguments are different with other marker options; it is not simply `i % nrow`, but we will have to store the original indices of `iconUrl` somewhere (e.g. in the current PR, I'm using `iconGroup`), and index the deduped `iconUrl`.
  - For efficiency, we need to dedup all four URLs `'iconUrl', 'iconRetinaUrl', 'shadowUrl', 'shadowRetinaUrl'`, and store the original indices somewhere;
  - For the current PR, I did not dedup them separately, but was only looking at the unique combinations of the four URLs (each unique combination was base64 encoded); this is less efficient than base64 encoding them separately and store four vectors of indices;

I feel a little lost here. Perhaps we can chat tomorrow.